File: plugins/lhalf/Keyrsla.py
from lhalf import *
from data_load import *
from matplotlib import pyplot as plt
import seaborn as sns
import numpy as np


#%%
datadir = '/Volumes/Bengal/Datasets/Jasper'
Y, S_ref, A_ref, n_row, n_col, n_end, n_band = load_data(datadir)
S = np.ones((n_col*n_row,n_end))*0.25
X = np.copy(Y)
q = 0.5
delta = 10
h = 1
max_iter = 5000
verbose = False
Ae = np.load('../../drasl/initial_endmembers.npy')
if verbose:
    A,S,J,SAD,t = lhalf(A_ref,Ae,S,X,delta,h,q,max_iter,verbose)
else:
    A,S = lhalf(A_ref,Ae,S,X,delta,h,q,max_iter,verbose)
spio.savemat('../../output/output.mat',{'A': A, 'S': S})


#%%
# plt.plot(A)
# plt.title(str(max_iter) + " iterations")
# plt.show()
# plt.hist(np.sum(S,axis=1),100)
# plt.title("mean(sum(S)) = " + str(np.mean(np.sum(S,axis=1))))
# plt.show()
# plt.plot(np.diff(J))
# plt.title("max(diff(J)) = " + str(np.max(np.diff(J))))
# plt.show()
# plt.plot(SAD)
# plt.title("min(SAD) = " + str(np.min(SAD)))
# plt.show()


#%%
# S_img = S.reshape((n_row,n_col,n_end))
# S_img = np.transpose(S_img,axes=(1,0,2))
# for abundance in range(S_img.shape[-1]):
#     plt.imshow(S_img[...,abundance])
#     plt.show()
#
# S_ref_img = np.transpose(S_ref,axes=(1,2,0))
# for abundance in range(S_ref_img.shape[-1]):
#     plt.imshow(S_ref_img[...,abundance])
#     plt.show()

File: matlabplugins/lhalf/Keyrsla.m
load('/Volumes/Music/Datasets/Samson/Mat/samson_1.mat')
load('/Volumes/Music/Datasets/Samson/GroundTruth/end3.mat')
nEnd = 3;
X = Y;
Aref = M;
Sref = A;
Ae = VCA(X,'Endmembers',nEnd);
%[Ae,~] = EIA_ATGP(X,nEnd);
%Ae = Aman;
Ae = max(Ae,0);

[sad_m,idx_org_m,idx_hat_m,sad_k,s0] = calcSAD2(Aref,Ae)
S = (ones(nEnd,nRow*nCol)*0.33).';
q = 0.5;
delta = 15;
h = 0.6;
maxIter = 5000;
calcMisc = 1;
tic;
[A,S,J,SAD,t]=lhalf(Aref,Ae,S,X,q,delta,h,maxIter,calcMisc);
[sad_m,idx_org_m,idx_hat_m,sad_k,s0] = calcSAD2(A,Aref)
size(A)
toc
figure(2);
display_unmix;

figure(3);
clear Simg;
Simg(:,1) = S(:,1)./max(S(:,1));
Simg(:,2) = S(:,2)./max(S(:,2));
Simg(:,3) = S(:,3)./max(S(:,3));
Simg = reshape (Simg(:,[1 2 3]), [nRow nCol nEnd]);
imshow(Simg)
caxis('auto')

save('/Volumes/Music/Unmixing/Python/lhalf/samsonVCA.mat', '-v7')
